lib/spack/spack/util/path.py

```diff
@@ -233,7 +249,9 @@ def add_padding(path, length):
     return os.path.join(path, padding)
 
 
-def canonicalize_path(path, default_wd=None):
+def canonicalize_path(
+    path, default_wd=None, env: Optional[FrozenEnvironment] = INITIAL_ENVIRONMENT
```

Comment:
```suggestion
    path, default_wd=None, env: Union[FrozenEnvironment, contextlib.AbstractContextManager] = spack.util.lang.nullcontext()
```
As currently implemented, INITIAL_ENVIRONMENT freezes things before Spack starts, but really it should be unfrozen until the build process starts so that Spack-managed environment variables can be used

Reply:
I'd simplify this further:

```suggestion
    path, default_wd=None, env: Optional[Dict] = None
```